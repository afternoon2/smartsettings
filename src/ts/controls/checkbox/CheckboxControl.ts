import cuid from 'cuid';

import { Control } from '../control/Control';
import { InternalState, Listener, ListenerUpdate, CheckboxControlOptions } from '../../types';

import Styles from '../../../sass/checkbox.sass';

export type CheckboxControlProps = {
  id: string,
  options: CheckboxControlOptions,
  parentElement: HTMLElement,
  sectionListener?: Listener,
  panelListener?: Listener,
};

export class CheckboxControl extends Control {
  public controlElement: HTMLInputElement;
  public readonly displayType = 'checkbox';

  protected static template = (state: InternalState): string => {
    const id: string = cuid();
    return `<label
      class="${Styles.checkbox}" for="${id}">
        <input
          id="${id}"
          type="checkbox"
          title="${state.name}"
          ${state.checked ? 'checked' : ''}
          ${state.disabled ? 'disabled' : ''}
          ${state.readonly ? 'readonly="true"' : ''}
        />
      </label>
    `;
  }

  constructor(props: CheckboxControlProps) {
    super({
      ...props,
      template: CheckboxControl.template,
    });
    this.controlElement = this.element.querySelector('[type="checkbox"]') as HTMLInputElement;
    this.listeners.set('checked', this.onCheck);
    this.bindActionListeners();
  }

  get checked(): boolean {
    return this.state.checked as boolean;
  }

  check() {
    this.state.checked = true;
  }

  uncheck() {
    this.state.checked = false;
  }

  toggle() {
    this.state.checked = !this.state.checked;
  }

  private onCheck: Listener = (update: ListenerUpdate) => {
    if (update.value === true) {
      this.controlElement.setAttribute('checked', 'true');
    } else {
      this.controlElement.removeAttribute('checked');
    }
  }

  private bindActionListeners() {
    this.controlElement.addEventListener('change', (event: Event) => {
      event.preventDefault();
      this.toggle();
    });
  }
}
