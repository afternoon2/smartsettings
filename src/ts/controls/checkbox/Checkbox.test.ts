import { CheckboxControl } from './CheckboxControl';

let checkbox: CheckboxControl;
let userListener: any;

describe('Button control', () => {
  beforeEach(() => {
    userListener = jest.fn();
    checkbox = new CheckboxControl({
      id: 'id',
      name: 'Button',
      options: {
        checked: false,
        checkboxId: 'checkboxId',
      },
      parentElement: document.body,
      userListener,
    });
  });

  test('If it matches the snapshot', () => {
    expect(checkbox.element).toMatchSnapshot();
  });

  test('Disable', () => {
    checkbox.disable();
    expect(checkbox.element).toMatchSnapshot();
  });

  test('Enable', () => {
    checkbox.disable();
    checkbox.enable();
    expect(checkbox.controlElement.hasAttribute('disabled')).toBe(false);
    expect(checkbox.element).toMatchSnapshot();
  });

  test('Check and uncheck', () => {
    checkbox.check();
    expect(checkbox.checked).toBe(true);
    checkbox.uncheck();
    expect(checkbox.checked).toBe(false);
  });

  test('Invoke onClick and userListener callback', () => {
    checkbox.check();
    expect(userListener).toHaveBeenCalledTimes(1);
  });
});
