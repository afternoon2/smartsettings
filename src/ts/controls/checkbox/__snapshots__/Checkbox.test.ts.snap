// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button control Disable 1`] = `
<div
  class="control disabled"
>
  <main
    class="control__body"
    id="id"
  >
    <p
      class="control__name"
      title="Button"
    >
      Button
    </p>
    <div
      class="control__content"
    >
      
    
      <label
        class="checkbox"
        for="checkboxId"
      >
        
      
        <input
          disabled="true"
          id="checkboxId"
          title="Button"
          type="checkbox"
        />
        
    
      </label>
      
  
    </div>
  </main>
</div>
`;

exports[`Button control Enable 1`] = `
<div
  class="control"
>
  <main
    class="control__body"
    id="id"
  >
    <p
      class="control__name"
      title="Button"
    >
      Button
    </p>
    <div
      class="control__content"
    >
      
    
      <label
        class="checkbox"
        for="checkboxId"
      >
        
      
        <input
          id="checkboxId"
          title="Button"
          type="checkbox"
        />
        
    
      </label>
      
  
    </div>
  </main>
</div>
`;

exports[`Button control If it matches the snapshot 1`] = `
<div
  class="control"
>
  <main
    class="control__body"
    id="id"
  >
    <p
      class="control__name"
      title="Button"
    >
      Button
    </p>
    <div
      class="control__content"
    >
      
    
      <label
        class="checkbox"
        for="checkboxId"
      >
        
      
        <input
          id="checkboxId"
          title="Button"
          type="checkbox"
        />
        
    
      </label>
      
  
    </div>
  </main>
</div>
`;
